ticket_upgrade_prediction/models/neural_net_torch/train.py

```diff
@@ -0,0 +1,260 @@
+            return Adam(self.model.parameters(), lr=learning_rate)
+        elif optimizer_name.lower() == "sgd":
+            return SGD(self.model.parameters(), lr=learning_rate)
+        else:
+            raise ValueError("Wrong type of optimizer provided!")
+
+    def _one_batch(self, X: torch.Tensor, y: torch.Tensor) -> None:
+        X.to(self.device)
+        y.to(self.device)
+
+        preds_logit = self.model(X.float())
+        loss = self.criterion(preds_logit.float(), y.float())
+
+        self.optimizer.zero_grad()
+        loss.backward()
+        self.optimizer.step()
+
+    @mlflow_run_start_handle
+    def fit(self, mlflow_run_name: Optional[int] = None) -> None:
+        to_mlflow = True if mlflow_run_name else False
+
+        train_dataset = UpgradeDataset(
+            X=self.dataset.X_train, y=self.dataset.y_train.values
+        )
+        train_loader = DataLoader(
+            dataset=train_dataset, batch_size=self.batch_size, shuffle=True
+        )
+
+        if to_mlflow:
+            mlflow.log_params(self.hparams.to_dict())
+
+        for epoch in tqdm(range(self.epochs), desc="Epoch"):
+            for X, y in tqdm(iter(train_loader), desc="Batch"):
+                self._one_batch(X=X, y=y)
+
+            with torch.no_grad():
+                self._evaluate(to_mlflow=to_mlflow, epoch=epoch)
+
+    def _evaluate(self, to_mlflow: bool, epoch: int) -> None:
+        evaluator = Evaluator(
+            model=self.model, X=self.dataset.X_test, y=self.dataset.y_test
+        )
+        metrics = evaluator.get_all_metrics(epoch=epoch, to_mlflow=to_mlflow)
+        # _ = evaluator.plot_all_plots(
+        #     save_path=self.plot_save_path, to_mlflow=to_mlflow
+        # )
+        logger.info(metrics)
+        self.training_results.append(metrics)
+
+    def get_results(self) -> list[Metrics]:
+        return self.training_results
+
+
+class NeuralNetHyperopt:
+    def __init__(
+        self,
+        data: pd.DataFrame,
+        hyper_params: dict,
+        n_splits: int = 5,
+        scaler: Any = StandardScaler,
+        y_col: str = "UPGRADED_FLAG",
+        train_size: float = 0.75,
+        per_fold_epoch: int = 10,
+        batch_size: int = 16,
+    ) -> None:
+        self.n_splits = n_splits
+        self.data = data
+        self.scaler = scaler
+        self.per_fold_epoch = per_fold_epoch
+        self.metrics = []
+        self.y_col = y_col
+        self.batch_size = batch_size
+        self.train_size = train_size
+        self.hyper_params_combinations = self._get_params_combinations(
+            hyper_params=hyper_params
+        )
+
+    def _get_params_combinations(self, hyper_params) -> list[dict]:
+        keys, values = zip(*hyper_params.items())
+        return [dict(zip(keys, v)) for v in itertools.product(*values)]
+
+    def _one_hparam_combination(self, hparams: dict, to_mlflow: bool):
+        s_kfold = StratifiedKFold(n_splits=self.n_splits)
```

Comment:
i believe it would be better if we could choose if we want stratified or non stratified k fold, then assign it to a var like k_fold and iterate over it

Reply:
Do we really need non stratified kfold?